x-pack/plugins/timelines/public/components/t_grid/helpers.tsx

```diff
@@ -87,27 +93,39 @@ const checkIfFieldTypeIsNested = (field: string, browserFields: BrowserFields) =
 const buildQueryMatch = (
   dataProvider: DataProvider | DataProvidersAnd,
   browserFields: BrowserFields
-) =>
-  `${dataProvider.excluded ? 'NOT ' : ''}${
-    dataProvider.queryMatch.operator !== EXISTS_OPERATOR &&
-    dataProvider.type !== DataProviderType.template
-      ? checkIfFieldTypeIsNested(dataProvider.queryMatch.field, browserFields)
-        ? convertNestedFieldToQuery(
-            dataProvider.queryMatch.field,
-            dataProvider.queryMatch.value,
-            browserFields
-          )
-        : checkIfFieldTypeIsDate(dataProvider.queryMatch.field, browserFields)
-        ? convertDateFieldToQuery(dataProvider.queryMatch.field, dataProvider.queryMatch.value)
-        : `${dataProvider.queryMatch.field} : ${
-            isNumber(dataProvider.queryMatch.value)
-              ? dataProvider.queryMatch.value
-              : escapeQueryValue(dataProvider.queryMatch.value)
-          }`
-      : checkIfFieldTypeIsNested(dataProvider.queryMatch.field, browserFields)
-      ? convertNestedFieldToExistQuery(dataProvider.queryMatch.field, browserFields)
-      : `${dataProvider.queryMatch.field} ${EXISTS_OPERATOR}`
-  }`.trim();
+) => {
+  const {
+    excluded,
+    type,
+    queryMatch: { field, operator, value },
+  } = dataProvider;
+
+  const isFieldTypeNested = checkIfFieldTypeIsNested(field, browserFields);
+  let queryString = excluded ? 'NOT ' : '';
+
+  switch (operator) {
+    case EXISTS_OPERATOR:
+      queryString =
+        queryString + buildEXISTSQueryMatch({ browserFields, field, isFieldTypeNested });
+      break;
+    case IS_OPERATOR:
+      if (!isStringOrNumberArray(value)) {
+        queryString =
+          queryString + type !== DataProviderType.template
+            ? buildISQueryMatch({ browserFields, field, isFieldTypeNested, value })
+            : buildEXISTSQueryMatch({ browserFields, field, isFieldTypeNested });
+      }
+      break;
+    case IS_ONE_OF_OPERATOR:
+      if (isStringOrNumberArray(value)) {
+        queryString =
+          queryString + buildISONEOFQueryMatch({ browserFields, field, isFieldTypeNested, value });
+      }
+      break;
```

Comment:
Consider using something like the `assertUnreachable` utility function in `x-pack/plugins/timelines/common/utility_types.ts` to add a `default` case that ensures this `switch` is exhaustive (at compile time), per the details in https://stackoverflow.com/questions/39419170/how-do-i-check-that-a-switch-block-is-exhaustive-in-typescript

Reply:
thats pretty cool, thanks for pointing it out